File: crates/mogeung-ui/src/diff.rs
//! Diff presentation: highlighting, word-level diffing, and side-by-side
//! pairing. Roadmap `R-D4`, `R-D5`, `R-D6`.
//!
//! All of it is pure — text in, spans out — so the interesting parts are
//! testable without a window, which is not true of anything that touches egui.
//!
//! **The highlighter is a tokenizer, not a parser.** No tree-sitter, no
//! grammars, no per-language configuration: it recognises strings, comments,
//! numbers and a shared keyword set. It will mis-colour things. That is an
//! acceptable trade for a reading aid with no dependencies and no build step;
//! if it ever needs to be *correct* rather than *helpful*, it should be
//! replaced wholesale rather than patched toward accuracy.

/// What a token is, for colouring.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Tok {
    Plain,
    Keyword,
    Str,
    Comment,
    Number,
    Type,
}

/// Keywords shared across the languages this tool actually sees. Deliberately
/// one flat set: per-language tables would be more correct and would need
/// language detection, which is a bigger commitment than a reading aid earns.
const KEYWORDS: &[&str] = &[
    "fn", "let", "mut", "pub", "use", "mod", "impl", "trait", "struct", "enum", "match", "if",
    "else", "for", "while", "loop", "return", "async", "await", "move", "ref", "where", "const",
    "static", "unsafe", "dyn", "as", "in", "self", "super", "crate", "type", "def", "class",
    "import", "from", "function", "var", "new", "this", "export", "default", "public", "private",
    "protected", "extends", "implements", "interface", "package", "func", "go", "defer", "chan",
    "range", "map", "nil", "None", "True", "False", "true", "false", "null", "undefined", "try",
    "catch", "except", "finally", "raise", "throw", "with", "yield", "lambda", "elif", "not",
    "and", "or", "is", "del", "pass", "break", "continue", "switch", "case", "do", "end",
];

fn is_ident(c: char) -> bool {
    c.is_alphanumeric() || c == '_'
}

/// Split one line of code into coloured spans.
///
/// Line-local: a string or block comment spanning lines is not tracked, because
/// a diff hunk routinely starts in the middle of one and there is no reliable
/// way to know. Mis-colouring a fragment is better than mis-colouring the rest
/// of the file.
pub fn highlight(line: &str) -> Vec<(Tok, String)> {
    let mut out: Vec<(Tok, String)> = Vec::new();
    let chars: Vec<char> = line.chars().collect();
    let mut i = 0;

    let push = |out: &mut Vec<(Tok, String)>, tok: Tok, s: String| {
        if s.is_empty() {
            return;
        }
        // Merge with the previous span when it has the same colour, so the
        // renderer emits far fewer widgets.
        if let Some((last_tok, last)) = out.last_mut() {
            if *last_tok == tok {
                last.push_str(&s);
                return;
            }
        }
        out.push((tok, s));
    };

    while i < chars.len() {
        let c = chars[i];

        // Comments run to end of line.
        let rest: String = chars[i..].iter().collect();
        if rest.starts_with("//") || rest.starts_with('#') || rest.starts_with("--") {
            push(&mut out, Tok::Comment, rest);
            break;
        }

        // Strings, with backslash escapes.
        if c == '"' || c == '\'' || c == '`' {
            let quote = c;
            let mut s = String::from(c);
            i += 1;
            while i < chars.len() {
                let d = chars[i];
                s.push(d);
                i += 1;
                if d == '\\' && i < chars.len() {
                    s.push(chars[i]);
                    i += 1;
                    continue;
                }
                if d == quote {
                    break;
                }
            }
            push(&mut out, Tok::Str, s);
            continue;
        }

        // Numbers, including hex and decimals.
        if c.is_ascii_digit() {
            let mut s = String::new();
            while i < chars.len()
                && (chars[i].is_ascii_alphanumeric() || chars[i] == '.' || chars[i] == '_')
            {
                s.push(chars[i]);
                i += 1;
            }
            push(&mut out, Tok::Number, s);
            continue;
        }

        // Identifiers: keyword, type-ish (leading capital), or plain.
        if is_ident(c) {
            let mut s = String::new();
            while i < chars.len() && is_ident(chars[i]) {
                s.push(chars[i]);
                i += 1;
            }
            let tok = if KEYWORDS.contains(&s.as_str()) {
                Tok::Keyword
            } else if s.chars().next().map(char::is_uppercase).unwrap_or(false) {
                Tok::Type
            } else {
                Tok::Plain
            };
            push(&mut out, tok, s);
            continue;
        }

        push(&mut out, Tok::Plain, c.to_string());
        i += 1;
    }
    out
}

/// A run of characters shared or not shared between two lines.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Span {
    pub text: String,
    pub changed: bool,
}

/// Word-level diff between a removed line and its added counterpart. `R-D5`.
///
/// Common prefix and suffix are found on *word* boundaries, and everything
/// between is marked changed. That is not a minimal edit script — a proper
/// Myers diff would highlight less — but it is O(n), never produces the
/// confetti that character-level diffing does on reformatted code, and answers
/// the only question being asked: *which part of this line actually moved?*
pub fn word_diff(old: &str, new: &str) -> (Vec<Span>, Vec<Span>) {
    // The leading `-`/`+` is a diff marker, not content. Including it would
    // make every pair differ at position 0, so the common-prefix scan would
    // stop immediately and light up the entire line — exactly the noise this
    // function exists to remove.
    let (old_sign, old_body) = split_marker(old);
    let (new_sign, new_body) = split_marker(new);

    let a = split_words(old_body);
    let b = split_words(new_body);

    let mut pre = 0;
    while pre < a.len() && pre < b.len() && a[pre] == b[pre] {
        pre += 1;
    }
    let mut suf = 0;
    while suf < a.len() - pre && suf < b.len() - pre && a[a.len() - 1 - suf] == b[b.len() - 1 - suf]
    {
        suf += 1;
    }

    let build = |sign: &str, words: &[&str]| -> Vec<Span> {
        let mut out: Vec<Span> = Vec::new();
        let mut add = |text: String, changed: bool| {
            if text.is_empty() {
                return;
            }
            if let Some(last) = out.last_mut() {
                if last.changed == changed {
                    last.text.push_str(&text);
                    return;
                }
            }
            out.push(Span { text, changed });
        };
        add(sign.to_string(), false);
        add(words[..pre].concat(), false);
        add(words[pre..words.len() - suf].concat(), true);
        add(words[words.len() - suf..].concat(), false);
        out
    };
    (build(old_sign, &a), build(new_sign, &b))
}

/// Split a diff line into its leading marker and its body.
fn split_marker(line: &str) -> (&str, &str) {
    match line.chars().next() {
        Some('+') | Some('-') | Some(' ') => line.split_at(1),
        _ => ("", line),
    }
}

/// Split into words, keeping separators as their own units so reassembly is
/// lossless.
fn split_words(s: &str) -> Vec<&str> {
    let mut out = Vec::new();
    let mut start = 0;
    let mut prev_ident = false;
    for (i, c) in s.char_indices() {
        let ident = is_ident(c);
        if i > start && ident != prev_ident {
            out.push(&s[start..i]);
            start = i;
        }
        prev_ident = ident;
    }
    if start < s.len() {
        out.push(&s[start..]);
    }
    out
}

/// One row of a side-by-side view. `R-D6`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Row {
    pub left: Option<String>,
    pub right: Option<String>,
}

/// Pair up a unified hunk's lines into left/right rows.
///
/// Runs of removals and additions are zipped so a modified line sits opposite
/// the line it replaced — which is what makes side-by-side worth having, and
/// what enables the word diff on that pair. Unmatched leftovers get a blank
/// on the other side.
pub fn side_by_side(lines: &[String]) -> Vec<Row> {
    let mut rows = Vec::new();
    let mut dels: Vec<String> = Vec::new();
    let mut adds: Vec<String> = Vec::new();

    fn flush(dels: &mut Vec<String>, adds: &mut Vec<String>, rows: &mut Vec<Row>) {
        let n = dels.len().max(adds.len());
        for i in 0..n {
            rows.push(Row {
                left: dels.get(i).cloned(),
                right: adds.get(i).cloned(),
            });
        }
        dels.clear();
        adds.clear();
    }

    for l in lines {
        match l.chars().next() {
            Some('-') => dels.push(l.clone()),
            Some('+') => adds.push(l.clone()),
            _ => {
                flush(&mut dels, &mut adds, &mut rows);
                rows.push(Row {
                    left: Some(l.clone()),
                    right: Some(l.clone()),
                });
            }
        }
    }
    flush(&mut dels, &mut adds, &mut rows);
    rows
}

#[cfg(test)]
mod tests {
    use super::*;

    fn toks(line: &str) -> Vec<Tok> {
        highlight(line).into_iter().map(|(t, _)| t).collect()
    }

    #[test]
    fn highlighting_is_lossless() {
        // The single property that matters: colouring must never change the
        // text. A renderer that drops or duplicates characters is worse than
        // no highlighting at all.
        for line in [
            "let x = \"hi\"; // note",
            "fn main() { let n = 0xFF_u8; }",
            "  # python comment",
            "s = 'it\\'s escaped'",
            "",
            "   ",
            "→ unicode ✓ and emoji 🎉",
            "let s = \"unterminated",
        ] {
            let rebuilt: String = highlight(line).into_iter().map(|(_, s)| s).collect();
            assert_eq!(rebuilt, line, "highlight() altered the text");
        }
    }

    #[test]
    fn recognises_the_obvious_tokens() {
        assert!(toks("let x = 1").contains(&Tok::Keyword));
        assert!(toks("x = \"str\"").contains(&Tok::Str));
        assert!(toks("x = 42").contains(&Tok::Number));
        assert!(toks("// hi").contains(&Tok::Comment));
        assert!(toks("let s = Session::new()").contains(&Tok::Type));
    }

    #[test]
    fn a_hash_inside_a_string_is_not_a_comment() {
        let got = highlight(r#"let url = "http://x/#frag";"#);
        // The `#` is inside the string span, so nothing is a comment.
        assert!(!got.iter().any(|(t, _)| *t == Tok::Comment));
    }

    #[test]
    fn word_diff_marks_only_what_moved() {
        let (old, new) = word_diff("-let timeout = 30;", "+let timeout = 60;");
        let changed_old: String = old.iter().filter(|s| s.changed).map(|s| s.text.clone()).collect();
        let changed_new: String = new.iter().filter(|s| s.changed).map(|s| s.text.clone()).collect();
        assert_eq!(changed_old, "30");
        assert_eq!(changed_new, "60");
    }

    #[test]
    fn word_diff_is_lossless_in_both_directions() {
        let cases = [
            ("-abc", "+abd"),
            ("-", "+"),
            ("-same", "+same"),
            ("-totally different", "+nothing alike here"),
            ("-a(b, c)", "+a(b, c, d)"),
        ];
        for (a, b) in cases {
            let (l, r) = word_diff(a, b);
            assert_eq!(l.iter().map(|s| s.text.clone()).collect::<String>(), a);
            assert_eq!(r.iter().map(|s| s.text.clone()).collect::<String>(), b);
        }
    }

    /// The marker is not content. A line moved verbatim from one side to the
    /// other has nothing highlighted.
    #[test]
    fn identical_bodies_have_nothing_changed() {
        let (l, r) = word_diff("-x", "+x");
        assert_eq!(l.iter().filter(|s| s.changed).count(), 0);
        assert_eq!(r.iter().filter(|s| s.changed).count(), 0);
    }

    /// Regression: the first implementation compared the `-`/`+` markers as
    /// content, so the common-prefix scan stopped at position 0 and every
    /// replacement highlighted the whole line — which is no better than not
    /// having a word diff at all.
    #[test]
    fn the_marker_never_widens_the_highlight() {
        let (old, new) = word_diff("-    self.timeout = 30;", "+    self.timeout = 60;");
        let changed_old: String =
            old.iter().filter(|s| s.changed).map(|s| s.text.clone()).collect();
        let changed_new: String =
            new.iter().filter(|s| s.changed).map(|s| s.text.clone()).collect();
        assert_eq!(changed_old, "30");
        assert_eq!(changed_new, "60");
    }

    #[test]
    fn side_by_side_pairs_a_replacement() {
        let lines: Vec<String> = vec![
            " ctx".into(),
            "-old one".into(),
            "+new one".into(),
            " tail".into(),
        ];
        let rows = side_by_side(&lines);
        assert_eq!(rows.len(), 3);
        assert_eq!(rows[1].left.as_deref(), Some("-old one"));
        assert_eq!(rows[1].right.as_deref(), Some("+new one"));
        // Context appears on both sides.
        assert_eq!(rows[0].left, rows[0].right);
    }

    #[test]
    fn side_by_side_handles_lopsided_runs() {
        let lines: Vec<String> = vec![
            "-a".into(),
            "-b".into(),
            "-c".into(),
            "+x".into(),
        ];
        let rows = side_by_side(&lines);
        assert_eq!(rows.len(), 3);
        assert_eq!(rows[0].right.as_deref(), Some("+x"));
        assert!(rows[1].right.is_none(), "extra removals need a blank right");
        assert!(rows[2].right.is_none());
    }

    #[test]
    fn side_by_side_preserves_every_line() {
        let lines: Vec<String> = vec![
            " a".into(), "-b".into(), "+c".into(), "-d".into(), " e".into(), "+f".into(),
        ];
        let rows = side_by_side(&lines);
        let mut seen: Vec<String> = Vec::new();
        for r in &rows {
            if let Some(l) = &r.left {
                if l.starts_with('-') || l.starts_with(' ') {
                    seen.push(l.clone());
                }
            }
            if let Some(x) = &r.right {
                if x.starts_with('+') {
                    seen.push(x.clone());
                }
            }
        }
        seen.sort();
        let mut want = lines.clone();
        want.sort();
        assert_eq!(seen, want, "side-by-side dropped or invented a line");
    }
}

File: desktop/src/lib/diff.ts
/**
 * Diff presentation: highlighting, word-level diffing, side-by-side pairing.
 * `R-D4`, `R-D5`, `R-D6`.
 *
 * A port of `crates/mogeung-ui/src/diff.rs`, function for function. All of it
 * is pure — text in, spans out — which is what let the original be tested
 * without a window and lets this one be tested without a DOM.
 *
 * **The highlighter is a tokenizer, not a parser.** No tree-sitter, no
 * grammars, no per-language configuration: it recognises strings, comments,
 * numbers and one shared keyword set. It will mis-colour things. That is an
 * acceptable trade for a reading aid with no dependencies and no build step —
 * and it is why this is not Shiki, which is in the tree for Monaco's sake and
 * would mean an async highlighter and a theme pipeline for the sake of colour
 * on a five-line hunk.
 */

export type Tok = "plain" | "keyword" | "string" | "comment" | "number" | "type";

/**
 * Keywords shared across the languages this tool actually sees. Deliberately
 * one flat set: per-language tables would be more correct and would need
 * language detection, which is a bigger commitment than a reading aid earns.
 */
const KEYWORDS = new Set([
  "fn", "let", "mut", "pub", "use", "mod", "impl", "trait", "struct", "enum", "match", "if",
  "else", "for", "while", "loop", "return", "async", "await", "move", "ref", "where", "const",
  "static", "unsafe", "dyn", "as", "in", "self", "super", "crate", "type", "def", "class",
  "import", "from", "function", "var", "new", "this", "export", "default", "public", "private",
  "protected", "extends", "implements", "interface", "package", "func", "go", "defer", "chan",
  "range", "map", "nil", "None", "True", "False", "true", "false", "null", "undefined", "try",
  "catch", "except", "finally", "raise", "throw", "with", "yield", "lambda", "elif", "not",
  "and", "or", "is", "del", "pass", "break", "continue", "switch", "case", "do", "end",
]);

function isIdent(c: string): boolean {
  return /[\p{L}\p{N}_]/u.test(c);
}

export interface Piece {
  tok: Tok;
  text: string;
}

/**
 * Split one line of code into coloured spans.
 *
 * Line-local: a string or block comment spanning lines is not tracked, because
 * a diff hunk routinely starts in the middle of one and there is no reliable
 * way to know. Mis-colouring a fragment beats mis-colouring the rest of a file.
 */
export function highlight(line: string): Piece[] {
  const out: Piece[] = [];
  const chars = [...line];
  let i = 0;

  const push = (tok: Tok, text: string) => {
    if (!text) return;
    const last = out[out.length - 1];
    // Merge same-coloured neighbours, so the renderer emits far fewer nodes.
    if (last && last.tok === tok) last.text += text;
    else out.push({ tok, text });
  };

  while (i < chars.length) {
    const c = chars[i];

    // Comments run to end of line. Checked on the char array directly — the
    // rest-of-line string is only built once, on a match, not per character.
    if ((c === "/" && chars[i + 1] === "/") || c === "#" || (c === "-" && chars[i + 1] === "-")) {
      push("comment", chars.slice(i).join(""));
      break;
    }

    // Strings, with backslash escapes.
    if (c === '"' || c === "'" || c === "`") {
      const quote = c;
      let s = c;
      i++;
      while (i < chars.length) {
        const d = chars[i];
        s += d;
        i++;
        if (d === "\\" && i < chars.length) {
          s += chars[i];
          i++;
          continue;
        }
        if (d === quote) break;
      }
      push("string", s);
      continue;
    }

    // Numbers, including hex and decimals.
    if (/[0-9]/.test(c)) {
      let s = "";
      while (i < chars.length && /[0-9a-zA-Z._]/.test(chars[i])) {
        s += chars[i];
        i++;
      }
      push("number", s);
      continue;
    }

    // Identifiers: keyword, type-ish (leading capital), or plain.
    if (isIdent(c)) {
      let s = "";
      while (i < chars.length && isIdent(chars[i])) {
        s += chars[i];
        i++;
      }
      push(KEYWORDS.has(s) ? "keyword" : /^\p{Lu}/u.test(s) ? "type" : "plain", s);
      continue;
    }

    push("plain", c);
    i++;
  }
  return out;
}

/** A run of characters shared or not shared between two lines. */
export interface Span {
  text: string;
  changed: boolean;
}

/** A diff line's leading marker and its body. */
function splitMarker(line: string): [string, string] {
  const c = line[0];
  return c === "+" || c === "-" || c === " " ? [line.slice(0, 1), line.slice(1)] : ["", line];
}

/**
 * Split into words, keeping separators as their own units so reassembly is
 * lossless.
 */
function splitWords(s: string): string[] {
  const out: string[] = [];
  let start = 0;
  let prevIdent = false;
  const chars = [...s];
  let at = 0;
  for (let i = 0; i < chars.length; i++) {
    const ident = isIdent(chars[i]);
    if (at > start && ident !== prevIdent) {
      out.push(s.slice(start, at));
      start = at;
    }
    prevIdent = ident;
    at += chars[i].length;
  }
  if (start < s.length) out.push(s.slice(start));
  return out;
}

/**
 * Word-level diff between a removed line and its added counterpart. `R-D5`.
 *
 * Common prefix and suffix are found on *word* boundaries and everything
 * between is marked changed. Not a minimal edit script — a proper Myers diff
 * would highlight less — but it is O(n), never produces the confetti that
 * character-level diffing makes of reformatted code, and answers the only
 * question being asked: which part of this line actually moved?
 */
export function wordDiff(oldLine: string, newLine: string): [Span[], Span[]] {
  // The leading `-`/`+` is a marker, not content. Including it would make every
  // pair differ at position 0, so the common-prefix scan would stop immediately
  // and light the whole line — exactly the noise this removes.
  const [oldSign, oldBody] = splitMarker(oldLine);
  const [newSign, newBody] = splitMarker(newLine);

  const a = splitWords(oldBody);
  const b = splitWords(newBody);

  let pre = 0;
  while (pre < a.length && pre < b.length && a[pre] === b[pre]) pre++;
  let suf = 0;
  while (
    suf < a.length - pre &&
    suf < b.length - pre &&
    a[a.length - 1 - suf] === b[b.length - 1 - suf]
  ) {
    suf++;
  }

  const build = (sign: string, words: string[]): Span[] => {
    const out: Span[] = [];
    const add = (text: string, changed: boolean) => {
      if (!text) return;
      const last = out[out.length - 1];
      if (last && last.changed === changed) last.text += text;
      else out.push({ text, changed });
    };
    add(sign, false);
    add(words.slice(0, pre).join(""), false);
    add(words.slice(pre, words.length - suf).join(""), true);
    add(words.slice(words.length - suf).join(""), false);
    return out;
  };

  return [build(oldSign, a), build(newSign, b)];
}

/** One row of a side-by-side view. `R-D6`. */
export interface Row {
  left: string | null;
  right: string | null;
  /**
   * The line's own number in each file, `null` where that side has no line —
   * and `null` on both when the hunk header could not be read.
   */
  leftNo: number | null;
  rightNo: number | null;
}

/** Where a hunk's two sides begin, in their own files. */
export interface HunkStart {
  old: number;
  new: number;
}

/**
 * The two starting line numbers of an `@@ -a,b +c,d @@` header.
 *
 * `null` rather than a default of `1` for anything it cannot read — a combined
 * merge header (`@@@`) is the case that turns up. A gutter of confidently wrong
 * numbers is worse than no gutter: you would go to the file and edit the wrong
 * line, and nothing on screen would have warned you.
 */
export function hunkStart(header: string): HunkStart | null {
  const m = /^@@ -(\d+)(?:,\d+)? \+(\d+)(?:,\d+)? @@/.exec(header);
  return m ? { old: Number(m[1]), new: Number(m[2]) } : null;
}

/**
 * Pair a unified hunk's lines into left/right rows, numbered.
 *
 * Runs of removals and additions are zipped so a modified line sits opposite
 * the line it replaced — which is what makes side-by-side worth having, and
 * what gives the word diff a pair to work on. Unmatched leftovers get a blank
 * opposite.
 *
 * The numbers come from `start`, walked down the hunk: each side advances only
 * on the rows that side actually has, which is what keeps a lopsided run — six
 * lines removed, two added — counting correctly on both.
 */
export function sideBySide(lines: readonly string[], start?: HunkStart | null): Row[] {
  const rows: Row[] = [];
  let dels: string[] = [];
  let adds: string[] = [];
  let old = start?.old ?? 0;
  let now = start?.new ?? 0;

  const push = (left: string | null, right: string | null) => {
    rows.push({
      left,
      right,
      leftNo: start && left !== null ? old++ : null,
      rightNo: start && right !== null ? now++ : null,
    });
  };

  const flush = () => {
    const n = Math.max(dels.length, adds.length);
    for (let i = 0; i < n; i++) push(dels[i] ?? null, adds[i] ?? null);
    dels = [];
    adds = [];
  };

  for (const l of lines) {
    const c = l[0];
    if (c === "-") dels.push(l);
    else if (c === "+") adds.push(l);
    else {
      flush();
      push(l, l);
    }
  }
  flush();
  return rows;
}

/**
 * The unified view's pairing: which added line, if any, a removed line
 * replaces — so a word diff can be drawn without switching to side-by-side.
 *
 * Same zip rule as [`sideBySide`], expressed as an index map because the
 * unified renderer walks the original array and needs to look sideways.
 */
export function pairs(lines: readonly string[]): Map<number, number> {
  const out = new Map<number, number>();
  let dels: number[] = [];
  let adds: number[] = [];
  const flush = () => {
    for (let i = 0; i < Math.min(dels.length, adds.length); i++) {
      out.set(dels[i], adds[i]);
      out.set(adds[i], dels[i]);
    }
    dels = [];
    adds = [];
  };
  lines.forEach((l, i) => {
    const c = l[0];
    if (c === "-") dels.push(i);
    else if (c === "+") adds.push(i);
    else flush();
  });
  flush();
  return out;
}
